lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,410 @@
+    GelfSerializeFailedInvalidFieldName, GelfSerializeFailedInvalidType,
+    GelfSerializeFailedMissingField,
+};
+use bytes::{BufMut, BytesMut};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::{log_schema, DataType},
+    event::Event,
+    event::LogEvent,
+    schema,
+};
+
+/// On GELF encoding behavior:
+///   Graylog has a relaxed parsing. They are much more lenient than the spec would
+///   suggest. We've elected to take a more strict approach to maintain backwards compatability
+///   in the event that we need to change the behavior to be more relaxed, so that prior versions
+///   of vector will still work.
+///   The exception is that if 'Additional fields' are found to be missing an underscore prefix and
+///   are otherwise valid field names, we prepend the underscore.
+
+/// Regex for matching valid field names. Must contain only word chars, periods and dashes.
+static VALID_FIELD: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[\w\.\-]*$").unwrap());
+
+static MISSING_FIELD_STR: &str = "LogEvent does not contain required field.";
+
+static INVALID_TYPE_STR: &str = "LogEvent contains a value with an invalid type.";
+
+static INVALID_FIELD_NAME_STR: &str = "LogEvent contains a value with an invalid type.";
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer {
+    conformed_log: Option<LogEvent>,
+}
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer {
+            conformed_log: None,
+        }
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        match event {
+            Event::Log(log) => serde_json::to_value(&log),
+            Event::Metric(_) | Event::Trace(_) => {
+                panic!("GELF Serializer does not support Metric or Trace events.")
```

Comment:
Do we know, in theory, when this could panic? When a metric or trace gets passed to the encoder? Could we wire up a configuration that actually triggers this, I'd think you'd need a sink that supports non-log events (as to not drop them) plus support for gelf encoding. Can `console` GELF encode?

Reply:
👍 just trying to understand when it might panic - didn't notice the supports -> to_json relationship

Reply:
I'll add a brief comment to that effect for this function, because yeah you kind of have to go to the definition of the Trait to see how it works.